### PR TITLE
Make WorkGroupMappingXCC and WorkGroupMappingXCCGroup as GPU kernel arguments

### PR DESCRIPTION
**Pull request by alex391a**
[LWPBLT-87](https://ontrack-internal.amd.com/browse/LWPBLT-87)

- Make `WGMXCC` and `WGMXCCG` as kernel Args.
1. `32-bit WGM control info = 10-bit WGMXCCG + 6-bit WGMXCC + 16-bit WGM`
2. Replace static divide/remainder with runtime divide/remainder.
3. Modify `scalarUInt32DivideAndRemainder` to enable `rReg=None`.
4. Add instruction `s_sext_i32_i16` to extend sign of 16-bit WGM.
- Enable `WGMXCCG=-1`
1. `WGMXCCG` set to `CU_Count` before passing the kernel args.
2. Add predicate to check `CU_Count % WGMXCC == 0` in runtime.

* hipblaslt lib size diff
`sudo du --max-depth=0 --block-size=M /build/release/hipblaslt-install/`
Before:
`13099M  hipBLASLt/build/release/hipblaslt-install/`
After:
`13031M  hipBLASLt/build/release/hipblaslt-install/`

### CODE REVIEW

File: tensilelite/Tensile/Source/lib/source/ContractionSolution.cpp [hunk 1]
```diff
@@ -830,10 +830,14 @@ namespace Tensile
...
         bool     gsuwgmrr = false; // initialized false
         int32_t  wgm      = param.wgm() != 0 ? param.wgm() : sizeMapping.workGroupMapping;
+        uint32_t wgmxcc   = 0;
+        uint32_t wgmxccg  = 0;
         const uint32_t mask16       = 0xFFFF;
         const uint32_t mask14       = 0x3FFF;
+        const uint32_t mask12       = 0xFFF;
         const uint32_t mask8        = 0xFF;
         uint32_t       internalArg0 = 0;
+        uint32_t        internalArg1 = 0;
```
Comment (jichangjichang) [inline tensilelite/Tensile/Source/lib/source/ContractionSolution.cpp]: not align, Have you ran the clang-format?

Comment (alex391a) [reply tensilelite/Tensile/Source/lib/source/ContractionSolution.cpp]: updated.

**Comment by alex391a** [comment]
[----------] Global test environment tear-down
[==========] 48206 tests from 13 test suites ran. (2633126 ms total)
[  PASSED  ] 48206 tests.
hipBLASLt version: 1000

========== 69 passed, 26 skipped, 38 warnings in 12536.94s (3:28:56) ============
py3: exit 0 (12537.22 seconds) /home/awang105/hipBLASLt/tensilelite> py.test -v --basetemp=/tmp/.tensile-tox/py3/tmp --junit-xml=/home/awang105/hipBLASLt/tensilelite/python_tests.xml --junit-prefix=py3 --color=yes -n 4 --prebuilt-client=/tmp/.tensile-tox/py3/client/0_Build/client/tensile_client Tensile/Tests -m common pid=3098255
  py3: OK (12697.15=setup[13.59]+cmd[0.43,145.91,12537.22] seconds)
  congratulations :) (12697.21 seconds)

File: tensilelite/Tensile/KernelWriterAssembly.py [hunk 2]
```diff
@@ -2005,6 +2016,15 @@ def graWorkGroup(self, kernel):
...
     # Blocked rows or columns
     # Do branch
+    
+    # Restore WGM
+    tmpSgpr = self.sgprPool.checkOut(2, preventOverflow=0)
+    module.add(SOrB32(dst=sgpr(tmpSgpr), src0=sgpr("WGM"), src1=hex(0xFFFFF000), comment=""))
+    module.add(SAndB32(dst=sgpr("WGM"), src0=sgpr("WGM"), src1=hex(0x00000FFF), comment=""))
+    module.add(SAndB32(dst=sgpr(tmpSgpr+1), src0=sgpr("WGM"), src1=hex(0x800), comment=""))
+    module.add(SCSelectB32(dst=sgpr("WGM"), src0=sgpr(tmpSgpr), src1=sgpr("WGM"), comment=""))
+    self.sgprPool.checkIn(tmpSgpr)
```
Comment (KKyang) [inline tensilelite/Tensile/KernelWriterAssembly.py]: WGM has negative values, can we keep it 16-bit so we can use native instruction to convert I16 to I32?

Comment (alex391a) [reply tensilelite/Tensile/KernelWriterAssembly.py]: Is it appropriate to set the range for XCC to be range(0, 128) and XCCG in range(1, 512), seems like we do not exceed these ranges?

Comment (alex391a) [reply tensilelite/Tensile/KernelWriterAssembly.py]: updated. 
add SSExtI16toI32

**Comment by alex391a** [comment]
[----------] Global test environment tear-down
[==========] 48206 tests from 13 test suites ran. (5979444 ms total)
[  PASSED  ] 48206 tests.
hipBLASLt version: 1000

========== 69 passed, 26 skipped, 38 warnings in 17526.54s (4:52:06) ===========
py3: exit 0 (17526.79 seconds) /home/awang105/hipBLASLt/tensilelite> py.test -v --basetemp=/tmp/.tensile-tox/py3/tmp --junit-xml=/home/awang105/hipBLASLt/tensilelite/python_tests.xml --junit-prefix=py3 --color=yes -n 4 --prebuilt-client=/tmp/.tensile-tox/py3/client/0_Build/client/tensile_client Tensile/Tests -m common pid=11998
py3: OK (17654.44=setup[9.12]+cmd[0.53,118.00,17526.79] seconds)
congratulations :) (17654.48 seconds)

**Review by KKyang — APPROVED** [review]

**Comment by alex391a** [comment]
Resolve conflict with #1035

**Comment by alex391a** [comment]
[----------] Global test environment tear-down
[==========] 48206 tests from 13 test suites ran. (3315133 ms total)
[  PASSED  ] 48206 tests.
hipBLASLt version: 1000

[33m========== [32m75 passed[0m, [33m[1m31 skipped[0m, [33m[1m48 warnings[0m[33m in 11056.36s (3:04:16)[0m[33m ===========[0m
py3: exit 0 (11056.53 seconds) /home/awang105/hipBLASLt/tensilelite> py.test -v --basetemp=/tmp/.tensile-tox/py3/tmp --junit-xml=/home/awang105/hipBLASLt/tensilelite/python_tests.xml --junit-prefix=py3 --color=yes -n 4 --prebuilt-client=/tmp/.tensile-tox/py3/client/0_Build/client/tensile_client Tensile/Tests -m common pid=733322
  py3: OK (11194.27=setup[11.90]+cmd[0.65,125.20,11056.53] seconds)
  congratulations :) (11194.31 seconds)

**Comment by alex391a** [comment]
Enable wgmxccg=-1, set wgmxccg to CU count in ContractionSolution.

[----------] Global test environment tear-down
[==========] 48206 tests from 13 test suites ran. (3665059 ms total)
[  PASSED  ] 48206 tests.
hipBLASLt version: 1000

[33m========== [32m75 passed[0m, [33m[1m31 skipped[0m, [33m[1m46 warnings[0m[33m in 11059.60s (3:04:19)[0m[33m ===========[0m
py3: exit 0 (11059.86 seconds) /home/awang105/hipBLASLt/tensilelite> py.test -v --basetemp=/tmp/.tensile-tox/py3/tmp --junit-xml=/home/awang105/hipBLASLt/tensilelite/python_tests.xml --junit-prefix=py3 --color=yes -n 4 --prebuilt-client=/tmp/.tensile-tox/py3/client/0_Build/client/tensile_client Tensile/Tests -m common pid=2517256
  py3: OK (11205.06=setup[11.56]+cmd[0.59,133.05,11059.86] seconds)
  congratulations :) (11205.12 seconds)

**Comment by hcman2** [comment]
For WGMXCCGroup=-1, why you choose to set the cu_count in ContractionSolution.cpp instead of in SolutionStructs.py? Is it possible to get the cu count in  SolutionStructs.py?

**Comment by alex391a** [comment]
> For WGMXCCGroup=-1, why you choose to set the cu_count in ContractionSolution.cpp instead of in SolutionStructs.py? Is it possible to get the cu count in SolutionStructs.py?

From @aazz44ss , if we need to keep CU_count private, we can't include it in the yaml file. Instead, we'll retrieve it at runtime and pass it as a kernel argument.

**Comment by hcman2** [comment]
does this patch impact the hipblaslt-bench performance? I remember that bench time includes this part.

File: tensilelite/Tensile/SolutionStructs.py [hunk 3]
```diff
@@ -2152,7 +2152,7 @@ def assignDerivedParameters(state):
 
     state["WorkGroupMappingXCC"] = abs(state["WorkGroupMappingXCC"])
 
-    if state["WorkGroupMappingXCCGroup"] % state["WorkGroupMappingXCC"] != 0:
+    if state["WorkGroupMappingXCCGroup"] != -1 and state["WorkGroupMappingXCCGroup"] % state["WorkGroupMappingXCC"] != 0:
```
Comment (jichangjichang) [inline tensilelite/Tensile/SolutionStructs.py]: Do we still need this checking? Should we check it in predicate stage (runtime)?

File: tensilelite/Tensile/Source/lib/include/Tensile/ContractionSolution.hpp [hunk 4]
```diff
@@ -145,8 +145,9 @@ namespace Tensile
 
         std::string customKernelName;
 
-        int workGroupMappingXCC = 1;
-        bool globalSplitUCoalesced = false;
+        int  workGroupMappingXCC                    = 1;
```
Comment (jichangjichang) [inline tensilelite/Tensile/Source/lib/include/Tensile/ContractionSolution.hpp]: default value is not zero?

File: tensilelite/Tensile/Source/lib/source/ContractionSolution.cpp [hunk 5]
```diff
@@ -986,13 +990,35 @@ namespace Tensile
...
 
+        if(internalArgsSupport.wgm && internalArgsSupport.version >= 1)
+        {
+            if(internalArgsSupport.version == 1)
+            {
+                internalArg1 = wgm;
+            }
+            else if(internalArgsSupport.version == 2)
+            {
+                wgmxcc = param.wgmxcc() > 0 ? param.wgmxcc() : sizeMapping.workGroupMappingXCC;
```
Comment (jichangjichang) [inline tensilelite/Tensile/Source/lib/source/ContractionSolution.cpp:1001]: what is the enabling condition? > 1 or > 0?

File: tensilelite/Tensile/Source/lib/source/ContractionSolution.cpp [hunk 6]
```diff
@@ -986,13 +990,35 @@ namespace Tensile
...
+                wgmxccg
+                    = param.wgmxccg() > 0 ? param.wgmxccg() : sizeMapping.workGroupMappingXCCGroup;
+                if(wgmxccg == -1)
```
Comment (jichangjichang) [inline tensilelite/Tensile/Source/lib/source/ContractionSolution.cpp]: should we calculate it when wgmxcc is enabled?

**Comment by alex391a** [comment]
Add predicate.
config:
```
        - WorkGroupMappingXCC: [8, 7]
        - WorkGroupMappingXCCGroup: [-1]
```
tuning output:
```
..._WGMXCC8_WGMXCCGn1,PASSED,...
..._WGMXCC8_WGMXCCGn1,PASSED,...
..._WGMXCC7_WGMXCCGn1,DID_NOT_SATISFY_ASSERTS,...
..._WGMXCC7_WGMXCCGn1,DID_NOT_SATISFY_ASSERTS,...
```

File: tensilelite/Tensile/Source/lib/include/Tensile/ContractionSolution.hpp (hunk 4 above)
Comment (alex391a) [reply tensilelite/Tensile/Source/lib/include/Tensile/ContractionSolution.hpp]: update

**Review by aazz44ss — DISMISSED** [review]
LGTM

**Comment by alex391a** [comment]
```
[----------] Global test environment tear-down
[==========] 48206 tests from 13 test suites ran. (3218213 ms total)
[  PASSED  ] 48206 tests.
hipBLASLt version: 1000
```

```
[33m========== [32m75 passed[0m, [33m[1m31 skipped[0m, [33m[1m46 warnings[0m[33m in 11034.52s (3:03:54)[0m[33m ===========[0m
py3: exit 0 (11034.78 seconds) /home/awang105/hipBLASLt/tensilelite> py.test -v --basetemp=/tmp/.tensile-tox/py3/tmp --junit-xml=/home/awang105/hipBLASLt/tensilelite/python_tests.xml --junit-prefix=py3 --color=yes -n 4 --prebuilt-client=/tmp/.tensile-tox/py3/client/0_Build/client/tensile_client Tensile/Tests -m common pid=51298
  py3: OK (11182.48=setup[12.62]+cmd[0.41,134.68,11034.78] seconds)
  congratulations :) (11182.52 seconds)
```

**Review by hcman2 — DISMISSED** [review]

**Review by KKyang — APPROVED** [review]

File: tensilelite/Tensile/Contractions.py [hunk 7]
```diff
@@ -498,6 +498,9 @@ def CompoundPredicates(cls, state, problemType):
             value = globalParameters['MinKForGSU']
             rv += [cls('GlobalSplitUCheckMinK', value=[value, state["GlobalSplitU"]])]
 
+        if ('WorkGroupMappingXCC' in state) and ('WorkGroupMappingXCCGroup' in state) and (state['WorkGroupMappingXCCGroup'] == -1):
```
Comment (jichangjichang) [inline tensilelite/Tensile/Contractions.py]: WorkgroupMappingXCCCheck is only necessary when WorkGroupMappingXCCGroup  is -1?

**Comment by alex391a** [comment]
[----------] Global test environment tear-down
[==========] 48206 tests from 13 test suites ran. (2247993 ms total)
[  PASSED  ] 48206 tests.
hipBLASLt version: 1000

[33m========== [32m75 passed[0m, [33m[1m31 skipped[0m, [33m[1m47 warnings[0m[33m in 11503.44s (3:11:43)[0m[33m ===========[0m
py3: exit 0 (11503.68 seconds) /home/awang105/hipBLASLt/tensilelite> py.test -v --basetemp=/tmp/.tensile-tox/py3/tmp --junit-xml=/home/awang105/hipBLASLt/tensilelite/python_tests.xml --junit-prefix=py3 --color=yes -n 4 --prebuilt-client=/tmp/.tensile-tox/py3/client/0_Build/client/tensile_client Tensile/Tests -m common pid=10629
  py3: OK (11639.19=setup[8.22]+cmd[0.63,126.65,11503.68] seconds)

File: tensilelite/Tensile/Source/lib/source/ContractionSolution.cpp [hunk 8]
```diff
@@ -986,13 +990,35 @@ namespace Tensile
...
 
+        if(internalArgsSupport.wgm && internalArgsSupport.version >= 1)
+        {
+            if(internalArgsSupport.version == 1)
+            {
+                internalArg1 = wgm;
+            }
+            else if(internalArgsSupport.version == 2)
+            {
+                wgmxcc = param.wgmxcc() > 1 ? param.wgmxcc() : sizeMapping.workGroupMappingXCC;
```
Comment (jichangjichang) [inline tensilelite/Tensile/Source/lib/source/ContractionSolution.cpp]: default value of param.wgmxcc() should be 0. wgmxcc should use the value of param.wgmxcc() if param.wgmxcc() > 0.

Comment (alex391a) [reply tensilelite/Tensile/Source/lib/source/ContractionSolution.cpp]: updated

**Review by jichangjichang — APPROVED** [review]
lgtm

**Review by aazz44ss — APPROVED** [review]
LGTM

**Comment by jichangjichang** [comment]
Please provide hipblaslt lib size diff with/WO this PR.

**Pull request merged**